network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -573,6 +573,19 @@ impl NetworkBehaviour for ConnectionPoolBehaviour {
         };
 
         self.addresses.mark_closed(address.clone());
+        // Notify handler about the connection is going to be shut down
+        self.actions
+            .push_back(NetworkBehaviourAction::NotifyHandler {
+                peer_id: *peer_id,
+                handler: NotifyHandler::Any,
```

Comment:
We can run into race conditions if we don't address the specific connection here with `NotifyHandler::One`. This is probably best changed in line 522